test/integration/producer_test.exs

```diff
@@ -0,0 +1,134 @@
+defmodule Mississippi.Integration.Producer.Test do
+  use ExUnit.Case
+
+  import Hammox
+
+  alias Mississippi.Consumer.AMQPDataConsumer.ExRabbitPoolConnection
+  alias Mississippi.Producer.EventsProducer
+
+  require Logger
+
+  @moduletag :integration
+
+  setup_all do
+    stub_with(MockAMQPConnection, ExRabbitPoolConnection)
+    Hammox.set_mox_global()
+
+    queue_count = System.unique_integer([:positive])
+    prefix = "mississippi_test_#{System.unique_integer()}_"
+    # We use the default exchange so that queues are binded using the routing key
+    exchange_name = ""
+
+    producer_options = [
+      amqp_producer_options: [host: "localhost"],
+      mississippi_config: [
+        queues: [events_exchange_name: exchange_name, total_count: queue_count, prefix: prefix]
+      ]
+    ]
+
+    start_consumer_pool!()
+
+    %{
+      producer: start_supervised!({Mississippi.Producer, producer_options}),
+      queue_count: queue_count,
+      queue_prefix: prefix
+    }
+  end
+
+  describe "Message sharding:" do
+    setup :create_sharding_key
+    setup :create_payload
+    setup :setup_message_handler
+
+    @tag :producer_message_sharding
+    test "Message is published on the correct queue according to the sharding key", %{
+      sharding_key: sharding_key,
+      payload: payload,
+      queue_count: queue_count,
+      queue_prefix: queue_prefix
+    } do
+      expected_queue_index = :erlang.phash2(sharding_key, queue_count)
+      expected_queue_name = "#{queue_prefix}#{expected_queue_index}"
+
+      _ = get_amqp_consumer_for(expected_queue_name)
+
+      EventsProducer.publish(payload, sharding_key: sharding_key)
+
+      assert_receive {^payload, headers, _timestamp}
+      assert :erlang.binary_to_term(headers["sharding_key"]) == sharding_key
+    end
+  end
+
+  describe "Message options:" do
+    setup :create_sharding_key
+    setup :create_payload
+    setup :setup_amqp_consumer
+
+    @tag :producer_message_options
+    test "Timestamp is added to the message if missing", %{
+      sharding_key: sharding_key,
+      payload: payload
+    } do
+      EventsProducer.publish(payload, sharding_key: sharding_key)
+
+      assert_receive {^payload, _headers, timestamp}
+      assert timestamp
+    end
+
+    @tag :producer_message_options
+    test "Timestamp is correctly included in the message if present", %{
+      sharding_key: sharding_key,
+      payload: payload
+    } do
+      timestamp = DateTime.to_unix(DateTime.utc_now())
```

Comment:
what happens if you pass and invalid timestamp or one with a wrong format?

Reply:
Timestamps are integer values, see [here](https://github.com/secomind/mississippi/blob/68aa2b08778b0a97d33ad9159a8d7482fac8f413/lib/producer/events_producer_options.ex#L65-L70). Anything other will return a validation error.
This allows the user to employ logical timestamps, if needed, rather than "clock" ones. But this design choice is not written in stone, what do you think?
Once we've decided, some unit tests for timestamps (and other fields, e.g. headers) can be added - in another PR

Reply:
sounds good